File: src/01.array/07.two_array_trans/01.566.ts
export function matrixReshape(
    mat: number[][],
    r: number,
    c: number
): number[][] {
    let m = mat.length;
    let n = mat[0].length;

    if (m * n !== r * c) {
        return mat;
    }

    let l = 0,
        k = 0;
    const res = [];

    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            if (k >= c) {
                k = 0;
                l++;
            }

            if (k === 0) {
                res.push([]);
            }

            res[l][k++] = mat[i][j];
        }
    }

    return res;
}

File: src/01.array/07.two_array_trans/01.566.go
package ArrayTwoTransform

func MatrixReshape(mat [][]int, r int, c int) [][]int {
	m := len(mat)
	n := len(mat[0])

	if m*n != r*c {
		return mat
	}

	l := 0
	k := 0
	res := [][]int{}

	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if k >= c {
				k = 0
				l++
			}

			if k == 0 {
				res = append(res, []int{})
			}

			res[l] = append(res[l], mat[i][j])
			k++
		}
	}

	return res
}
